design/baremetal-operator/bulk-set-bios-config.md

```diff
@@ -0,0 +1,215 @@
+        "CurrentValue": null,
+        "DisplayName": "SR-IOV Global Enable",
+        "DisplayOrder": 2331,
+        "HelpText": "Enables or disables the BIOS configuration of Single Root
+                     I/O Virtualization (SR-IOV) devices. Enable this feature
+                     if booting to a virtualization operating system that
+                     recognize SR-IOV devices. ",
+        "Hidden": false,
+        "Immutable": false,
+        "MenuPath": "./IntegratedDevicesRef",
+        "ReadOnly": false,
+        "ResetRequired": true,
+        "Type": "Enumeration",
+        "Value": [
+          {
+            "ValueDisplayName": "Enabled",
+            "ValueName": "Enabled"
+          },
+          {
+            "ValueDisplayName": "Disabled",
+            "ValueName": "Disabled"
+          }
+        ],
+        "WarningText": null,
+        "WriteOnly": false
+    }
+
+The cached BIOS registry from Ironic will also be made available in the
+BaremetalHost CRD. This can be a ConfigMap referenced by the
+`BaremetalHostStatus`. Since the BIOSRegistry can be quite large this should be
+stored gzip compressed so as to not exceed the 1Mb data limit in the ConfigMap.
+
+Using the BIOS settings and BIOS registry, the user will be able create a JSON
+list of bulk BIOS settings (aka “opaque blob”). This same JSON blob can be set
+on multiple machines to load and apply, effectively cloning all the settings
+from another machine. This user data will also be stored in a ConfigMap
+referenced in the BareMetalHostStatus. The BMO will send the data to Ironic via
+the clean-steps API without validating or interpreting it. Ironic will convert
+the JSON to a set of clean steps and run cleaning to configure the BIOS
+settings.
+
+### Implementation Details/Notes/Constraints
+
+One of the primary drivers of this design was the desire to not have the
+Installer validate the user inputted bulk BIOS settings as it requires vendor
+specific checks that are difficult to maintain. This proposal does not suggest
+validity checks in the installer or Ironic either - the input blob is converted
+to clean steps as-is.
+
+In order to avoid applying attributes that are ReadOnly when setting BIOS
+configuration, which will have no effect and will increase the time for BIOS
+configuration it is recommended that the user remove ReadOnly fields, as
+identified in the BIOS Registry, when building the JSON input. It’s not
+expected that the entire set read from Ironic will be passed down as:
+1. many of the values read from Ironic are ReadOnly
+2. setting ReadOnly values to their current settings increases the size of the
+input and may potentially affect the Redfish request to the BMC
+
+The BIOS Settings are changed during the node cleaning step which requires a
+reboot.  Manual clean occurs in the Preparing state of the BaremetalHost, the
+Host will re-enter this state from Ready/Available state whenever its config
+differs from the last one it stored.
```

Comment:
I think we've previously said we only wanted to apply BIOS settings when metal3 thinks it needs to reprovision the host. I can't tell from the way this is phrased if you're suggesting that we include BIOS settings in the triggers for reprovisioning, or if we would still only apply BIOS settings when a new image is going to be written.

Reply:
Yes, rewrote this to indicate that a change to BIOS settings will not require reprovisioning.  The settings will be written during the provisioning stage if the desired (bulk) settings don't match the current settings.